js/src/dom/data.js

```diff
@@ -11,57 +11,47 @@
  * ------------------------------------------------------------------------
  */
 
-const mapData = (() => {
-  const storeData = {}
-  let id = 1
-  return {
-    set(element, key, data) {
-      if (typeof element.bsKey === 'undefined') {
-        element.bsKey = {
-          key,
-          id
-        }
-        id++
-      }
+const elementMap = new Map()
 
-      storeData[element.bsKey.id] = data
-    },
-    get(element, key) {
-      if (!element || typeof element.bsKey === 'undefined') {
-        return null
-      }
-
-      const keyProperties = element.bsKey
-      if (keyProperties.key === key) {
-        return storeData[keyProperties.id]
-      }
+export default {
+  set(element, key, instance) {
+    if (!elementMap.has(element)) {
+      elementMap.set(element, new Map())
+    }
 
-      return null
-    },
-    delete(element, key) {
-      if (typeof element.bsKey === 'undefined') {
-        return
-      }
+    const instanceMap = elementMap.get(element)
 
-      const keyProperties = element.bsKey
-      if (keyProperties.key === key) {
-        delete storeData[keyProperties.id]
-        delete element.bsKey
-      }
+    // make it clear we only want one instance per element
+    // can be removed later when multiple key/instances are fine to be used
+    if (!instanceMap.has(key) && instanceMap.size !== 0) {
+      // eslint-disable-next-line no-console
+      console.error(`Notice: The design of Bootstrap currently discourages to use more than one instance per element. Bound instance(s): ${Array.from(instanceMap.keys()).join(', ')}.`)
```

Comment:
Now you can omit **s** from the instance(s) since there is only one instance bound.
```suggestion
      console.error(`Notice: The design of Bootstrap currently discourages to use more than one instance per element. Bound instance: ${Array.from(instanceMap.keys())[0]}.`)

```

Reply:
I think we should also change the `discourages` to `disallow` then and maybe remove `currently` as this isn't something that is likely to be addressed in the closer future.

Reply:
/CC @XhmikosR